src/main/scala/net/snowflake/spark/snowflake/io/SFInternalRDD.scala

```diff
@@ -0,0 +1,190 @@
+}
+
+private[io] class SFInternalRDD(
+                                        @transient val sqlContext: SQLContext,
+                                        @transient val params: MergedParameters,
+                                        @transient val sql: String,
+                                        @transient val jdbcWrapper: JDBCWrapper,
+                                        val format: SupportedFormat = SupportedFormat.CSV
+) extends RDD[String](sqlContext.sparkContext, Nil) with DataUnloader {
+
+  @transient override val log = LoggerFactory.getLogger(getClass)
+  @transient private val stageManager =
+    new SFInternalStage(false, jdbcWrapper, params)
+
+  @transient val tempStage = stageManager.setupStageArea()
+
+  @transient private val FILES_PER_PARTITION = 2
+
+  private val compress = if (params.sfCompress) "gzip" else "none"
+  private val parallel = params.parallelism
+
+  private[io] val rowCount = stageManager
+    .executeWithConnection({ c =>
+      setup(preStatements = Seq.empty,
+            sql = buildUnloadStmt(sql, s"@$tempStage", compress, None),
+            conn = c,
+            keepOpen = true)
+    })
+    .asInstanceOf[Long]
+
+  private val stageLocation = stageManager.stageLocation
+  private val awsID         = stageManager.awsId
+  private val awsKey        = stageManager.awsKey
+  private val awsToken      = stageManager.awsToken
+  private val masterKey     = stageManager.masterKey
+  private val is256         = stageManager.is256Encryption
+
+  private val azureSAS      = stageManager.azureSAS
+  private val azureAccount  = stageManager.azureAccountName
+  private val azureEndpoint = stageManager.azureEndpoint
+
+  private val stageType: StageInfo.StageType = stageManager.stageType
+
+  override def getPartitions: Array[Partition] = {
+
+    val encryptionMaterialsGrouped =
+      stageManager.getKeyIds.grouped(FILES_PER_PARTITION).toList
+
+    val partitions = new Array[Partition](encryptionMaterialsGrouped.length)
+
+    var i = 0
+
+    while (i < encryptionMaterialsGrouped.length) {
+      partitions(i) =
+        new SFInternalPartition(encryptionMaterialsGrouped(i), id, i)
+      i = i + 1
+    }
+    partitions
+  }
+
+  override def compute(thePartition: Partition,
+                       context: TaskContext): Iterator[String] = {
+
+    val mats   = thePartition.asInstanceOf[SFInternalPartition].srcFiles
+
+    val stringIterator = new SFRecordReader(format)
+    try {
+      mats.foreach {
+        case (file, queryId, smkId) =>
+          if (queryId != null) {
+            var stream: InputStream = null
+            val (bucketName, stagePath) =
+              extractBucketNameAndPath(stageLocation)
+            var stageFilePath = file
+            if (!stagePath.isEmpty) {
+              stageFilePath =
+                SnowflakeUtil.concatFilePathNames(stagePath, file, "/")
+            }
+
+
+            stageType match {
```

Comment:
My preference is to use polymorphism instead of a switch case here to differentiate between Azure and S3, but I'll leave the final decision up to you.